hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/UpsertPartitioner.java

```diff
@@ -367,9 +368,19 @@ public int getPartition(Object key) {
   /**
    * Obtains the average record size based on records written during previous commits. Used for estimating how many
    * records pack into one file.
+   * Respect user setting by following the precedence as below
+   * 1) if user sets a value, then use it as is
+   * 2) if user not setting it, infer from timeline commit metadata
+   * 3) if timeline is empty, use a default (current: 1024)
    */
   protected static long averageBytesPerRecord(HoodieTimeline commitTimeline, HoodieWriteConfig hoodieWriteConfig) {
+    long defaultAvgSize = Integer.parseInt(HoodieCompactionConfig.COPY_ON_WRITE_RECORD_SIZE_ESTIMATE.defaultValue());
     long avgSize = hoodieWriteConfig.getCopyOnWriteRecordSizeEstimate();
+
+    if (avgSize != defaultAvgSize) {
```

Comment:
The `HoodieCompactionConfig#build` already sets the defaults explicitly if it is not configured, so what's the problem here ?

Reply:
I see, you didn't what the moving average from the commit matadata, is the moving average not accurate or something else ?

Reply:
@danny0405 This pr makes  user-provided avg value first. We can't tell the value is default or user-provided here, so I suppose if the avg size is not default, it is user provided.

Reply:
Yeah, you need to explain why the explicit record size is always considered while ignoring the write stats from the commit metadata ? Can we elaborate to make the commit metadata more accurate here ?

Reply:
Hi @danny0405 , do you mean to add some comment to explain why we ignore the write stats from the commit metadata and how to set the value more accurate?

Reply:
I mean we should figure out why the write stats is not that accurate ?

Reply:
@danny0405  The calculation of average value is OK from my side.
@xushiyan could you explain the purpose of making user-provided first？Are these changes what you expected？